### ISSUE TITLE
Can't run Python Agent: ImportError: DLL load failed

### ISSUE THREAD

**Issue by TheAeryan** [Microsoft/malmo#736]
Hi! I have just installed Malmo from the pre-built version https://github.com/Microsoft/malmo/blob/master/doc/install_windows_manual.md and follow the instructions for a manual installation (except for the C# environment)(https://github.com/Microsoft/malmo/blob/master/doc/install_windows_manual.md).
The game (the minecraft window) runs perfectly and I have been able to run a C++ agent. However, when i try to run the first tutorial on python (python tutorial_1.py) I get the following error:
Traceback (most recent call last):
  File "tutorial_1.py", line 24, in <module>
    import MalmoPython
ImportError: DLL load failed: The specified module could not be found.

I have seen other threads about the same problem talking about CodeSynthesis XSD and how it had to be included in the Path. However, the tutorial mentioned above doesn't say a word about it.
I had also thought that maybe I had to use pip3 to install the python modules, but when I write in the command prompt _pip3 install malmo_ I get the following message:
Collecting malmo
  Could not find a version that satisfies the requirement malmo (from versions: )
No matching distribution found for malmo

Another thing I suspect could be causing this problem is that I have the _MalmoPython.lib_ file in the Python_Examples folder but I'm missing the  _MalmoPython.so_ file in the same folder.

My OS is Windows 10 and my python version is 3.7.0, the 64-bit version.

**Comment by okkhoy** [comment]
This could be due to Python version mismatch. The binaries are built for Python 3.6.
WRT _MalmoPython.lib_ vs _MalmoPython.so_, the `.lib` version is correct (since you are on Windows). 
`.so` files are for Linux.

**Comment by TheAeryan** [comment]
Thanks for your reply! I didn't know I only needed to have the .lib file instead of both versions on windows.
However, since I am using python 3.7.0 I don't think there's much difference between it and the 3.6.0 version.
I continued my search and I found out that, since I hadn't used the PowerShell script, I had some dependencies left, like CodeSynthesis XSD. The problem with the script is that when it gets to the Java installation part it just freezes and after a long time this error appears:
![problema_java](https://user-images.githubusercontent.com/41746312/43317841-c34c4040-919e-11e8-8fd9-cf6892eb3a3a.png)

I already have Java installed so I think the problem could maybe be solved by modifying the script so that it didn't try to install Java but I don't know PowerShell so I can't do it.

**Comment by rs4da** [comment]
So I actually had the same problem myself when performing my first installation. Your issue stems from the fact that the the ffmpeg download link in malmo_install.ps1 is broken. Just go to this link https://ffmpeg.zeranoe.com/builds/ and get the latest build for windows 10. That should hopefully fix your issue. From her just try to run tutorial_1.py. You should have all the dependencies already set up.

**Comment by TheAeryan** [comment]
rs4da, if I have understood it correctly, you think the problem could be solved by installing that version of ffmpeg. The thing is that, the first time I ran the script, it installed in my machine the version that is written in the script (the one you say it's broken) and the several times I ran it after the first time, it checked for the installation and didn't reinstall it (since it's already installed). But when it gets to the Java part, it gives me that error and, since there are other dependencies (such as CodeSynthesis XSD) left, I don't think it will work even if I install that version of ffmpeg, unless I get the other dependencies set up. Am I wrong?
The only thing that would make sense is if that wrong version of ffmpeg is preventing the script from installing Java/recognizing it is already installed, but as I said, I think I don't have all the dependencies set up yet.

**Comment by rs4da** [comment]
Hmm I'm not really sure what to do if that's the case (may have jumped the gun a little with my "miraculous" solution lol). Are you sure you are using the developer version of Java?

**Comment by rs4da** [comment]
Since you say you are able to run the minecraft client without problems the problem likely lies in your python as opposed to your Java, as suggested by okkhoy. If Java was broken minecraft would not launch.

**Comment by TheAeryan** [comment]
Pretty sure (if you mean the JDK), since I have done several projects in Java.
![version_java](https://user-images.githubusercontent.com/41746312/43326332-df2e31f0-91b8-11e8-82c4-c8ea3355fd2d.png)
I think I only need to make the script recognize I already have Java installed so that it can set up the remaining dependencies. The problem is that I don't know why it doesn't see it (I have the environmental variables like JAVA_HOME correctly set up). It could be, as you mentioned, that I have the wrong version of ffmpeg, but if it prevents the script from recognising it, it would be very odd.

**Comment by rs4da** [comment]
Once all the dependencies are set up you are good to go and don't need to use pip. As for ffmpeg, I'd give it a shot if I were you, maybe it'll fix it, who knows. If not, I'd try downgrading to python 3.6, although I agree with you in that there shouldn't be conflicts between 3.6 and 3.7. Best of luck!

**Comment by TheAeryan** [comment]
Ok, thanks! for now I'll try it and see how it goes. By the way, if I install the version you mentioned of ffmpeg, the script should see it is already installed although it's a different version from the standard one and not try to reinstall it, shouldn't it?

**Comment by rs4da** [comment]
In theory, yes. When you reinstall ffmpeg you should try running the Malmo code without using the install_malmo script again, since you should have all the dependencies set up at that point. Personally I got Malmo working by doing the manual dependency installation since the script never really worked for me. Lmk how it goes.

**Comment by okkhoy** [comment]
Based on my experience, as long as there is ffmpeg installed it is ok. 
As suggested by @rs4da pick up the latest stable version of ffmpeg from zeranoe website; it should work.

WRT the powershell script; I never tried it.

- You can carefully follow the install instructions in the manual installation and see if you have followed each step (set aside the powershell script, we wil not need it if we do the manual installation)
- Create a new virtual env with Python 3.6 and try running the examples; if this works, you are sure that it is due to Python version.

**Comment by TheAeryan** [comment]
Ok, I'll first try installing the latest version of ffmpeg and if that's not enough I'll downgrade from python 3.7 to 3.6.
I thought I needed to set up more dependencies than the ones that appear on the manual installation guide (for example, it doesn't mention CodeSynthesis XSD nor Git, but I think the script tries to set up those dependencies too).

**Comment by TheAeryan** [comment]
I have tried reinstalling the latest version of ffmpeg and it made no difference. However, I think I have finally (thanks God) found what the problem was. As mentioned in #725, Python 3.7 can't be used but the version 3.6 instead. It seems you **do** need to download (apart from setting up the relevant dependencies) the malmo modules using pip.
Currently, there's a version for Python 3.7 but only for Mac OS (I tried installing it on my windows machine and it didn't work) so it seems I'll have to downgrade or set up a virtual environment for Python 3.6.
The malmo libraries are in here: _https://pypi.org/project/malmo/#files_. I don't really know if I can make a request for a 3.7 version for Windows (as there is for Mac) but I suppose it would really help people since most of them will have the newest version of Python instead of the 3.6.

**Comment by TheAeryan** [comment]
I have now uninstalled version 3.7 and installed the 3.6 one (since I didn't really want to use a virtual environment) and it's working fine!
As I said in the previous post, you **do** need to install the python modules using pip, since they aren't automatically installed with the dependencies. Currently, the only 3.7 available version is for Mac, so it seems we'll need to wait for the creators to upgrade the modules to work with 3.7 or stick to using Python 3.6; it's fine to me, though.
Thank you very much to @rs4da and @okkhoy for your help. You really helped me a lot!
Issue is now closed.
